shell/shared/grain.js

```diff
@@ -511,6 +511,60 @@ if (Meteor.isClient) {
             currentGrainId + event.data.setPath);
       } else if (event.data.setTitle) {
         Session.set("grainFrameTitle", event.data.setTitle);
+      } else if (event.data.renderTemplate) {
+        // Request creation of a single-use template with a privileged API token.
+        // Why?  Apps should not be able to obtain capabilities to themselves
+        // directly, but often need a way to provide instructions to users to
+        // copy/paste with some privileged token contained within.  By
+        // providing this templating in the platform, we can ensure that the
+        // token is only visible to the shell's origin.
+        var call = event.data.renderTemplate;
+        check(call, Object);
+        var rpcId = call.rpcId;
+        check(rpcId, String);
+        var template = call.template;
+        check(template, String);
+        var assignment = undefined;
+        if (call.roleId) {
+          check(call.roleId, Match.Integer);
+          assignment = {roleId: call.roleId};
+        } else {
+          assignment = {none: null};
+        }
+        var petname = undefined;
+        if (call.petname) {
+          check(call.petname, String);
+          petname = call.petname;
+        } else {
+          petname = "template token";
+        }
+        // Tokens expire by default in 5 minutes from generation date
```

Comment:
Let's make sure to implement some sort of periodic ping from the template HTML that prevents the token from expiring while it's still open.